### ISSUE TITLE
Failure to column-wise-concatenate matrix with itself?

### ISSUE THREAD

**Issue by davidmrdavid** [josdejong/mathjs#1727]
The following snippet of code loads a matrix **A** from disk and attempts to perform a
column-wise concatenation of it with itself. The original dimensions of **A** are
[ 253120, 1 ] so I expect the output to be [ 253120, 2 ].

```JavaScript
const fs = require("fs");
const math = require("mathjs");

function main () {
    contents = fs.readFileSync("A.json");
    json = JSON.parse(contents);
    let A = math.DenseMatrix.fromJSON(json);
    console.log(A.size());
    let A2 = math.concat(A, A, 1); // This will fail
}

main();
```

Unfortunately, the script fails with the following error trace:

```
bash-4.2$ node --max-old-space-size=20000 reproducer1.js
[ 253120, 1 ]

/scratch_user/djusto/tester/node_modules/mathjs/lib/function/matrix/concat.js:91
            throw new _IndexError.IndexError(dim, prevDim + 1);
            ^
Error
/...a larger stack trace .../
bash-4.2$
```

What could this be? The error suggests a dimensionality error but that couldn't be since I'm concatenating a matrix with itself. Thanks!

Files:
[reproduce1.zip](https://github.com/josdejong/mathjs/files/4125211/reproduce1.zip)

**Comment by josdejong** [comment]
A DenseMatrix is serialized as nested arrays. The `data` in the A.json has a single dimension though, and no nesting, so the dimensions in `size` do not correspond with the number of dimensions mathjs expects in `data`. 

To see what data format is expected for DenseMatrix, run:

```js
const A = math.evaluate('[1; 2; 3]')
console.log(A.size()) // [ 3, 1 ]
console.log(A.toJSON())
// {
//  mathjs: 'DenseMatrix',
//  data: [ [ 1 ], [ 2 ], [ 3 ] ],
//  size: [ 3, 1 ],
//   datatype: undefined
// }
```

It would be good I think to improve `fromJSON` such that it does do a basic check to see whether `data` and `size` correspond with each other.

**Comment by josdejong** [comment]
I've added validation in commit 0c863f4e073c14f96d79e769518347b72d1aca3f. The error message itself could use some improvement too, but at least it's now giving an error when _creating_ the matrix and the provided size doesn't match the data.

Adding the validation made some unit tests fail: there where some issues in unit tests and a few internal algorithms, and there was a but in `DenseMatrix.map`. So that is really nice :+1:

**Comment by josdejong** [comment]
Fix is available now in `v6.6.0`
